apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,428 @@
+      this._setProfileConnected(this.Profiles.A2DP, evt.status);
+    });
+
+    adapter.addEventListener('scostatuschanged', (evt) => {
+      this._setProfileConnected(this.Profiles.SCO, evt.status);
+    });
+  },
+
+  /**
+   * BT APIv2: Watch 'onattributechanged' event from
+   * mozBluetooth.defaultAdapter for updating state information.
+   *
+   * @private
+   * @param  {Object} evt event object
+   */
+  _bluetoothAdapterHandler:
+    function bt__bluetoothAdapterHandler(evt) {
+      for (var i in evt.attrs) {
+        switch (evt.attrs[i]) {
+          case 'state':
+            if (this._adapter.state === 'enabled') {
+              this._isEnabled = true;
+              navigator.mozSettings.createLock()
+                .set({'bluetooth.enabled': true});
+              window.dispatchEvent(new CustomEvent('bluetooth-enabled'));
+              // this.icon && this.icon.update();
+            } else if (this._adapter.state === 'disabled') {
+              this._isEnabled = false;
+              navigator.mozSettings.createLock()
+                .set({'bluetooth.enabled': false});
+              window.dispatchEvent(new CustomEvent('bluetooth-disabled'));
+              // this.icon && this.icon.update();
+            }
+            break;
+          default:
+            break;
+        }
+      }
+  },
+
+  /**
+   * Watch 'onattributechanged' event from mozBluetooth for updating default
+   * adapter information.
+   *
+   * 'onattributechanged' event description:
+   * A handler to trigger when bluetooth manager's only property
+   * defaultAdapter has changed.
+   *
+   * @private
+   * @param  {Object} evt event object
+   */
+  _bluetoothManagerHandler:
+    function bt__bluetoothManagerHandler(resolve, evt) {
+      for (var i in evt.attrs) {
+        switch (evt.attrs[i]) {
+          case 'defaultAdapter':
+            this._debug('defaultAdapter changed.');
+            if (this._bluetooth.defaultAdapter) {
+              // Default adapter attribute change.
+              // Usually, it means that we reach new default adapter.
+              this._adapter = this._bluetooth.defaultAdapter;
+              this._isEnabled = (this._adapter.state === 'enabled');
+              this._stateChangeHandler =
+                this._bluetoothAdapterHandler.bind(this);
+              this._adapter.addEventListener('attributechanged',
+                this._stateChangeHandler);
+              this._updateProfileState(this._adapter);
+              if (typeof resolve === 'function') {
+                resolve(this._adapter);
+              }
+            }
+            break;
+          default:
+            break;
+        }
+      }
+  },
+
+  /**
+   * Called by external for re-use adapter.
+   * For defaultAdapter not ready case, we cached the Promise object
+   * therefore all caller would get the same Promise object.
+   *
+   * @public
+   */
+  getAdapter: function bt_getAdapter() {
```

Comment:
Given the current bluetooth API design, an adapter becomes available when gecko finds one and which is not triggered by gaia, so it makes less sense wrapping the function as an async call that returns a promise. Consider this case, if gecko fails to report an adapter, the async call gets stuck forever.
I would suggest simply returns the cached adapter. If it is null, just return null. All other functions should also be designed as event triggered.